src/crontab.c

```diff
@@ -479,16 +479,89 @@ static void delete_cmd(void) {
 		fprintf(stderr, "path too long\n");
 		exit(ERROR_EXIT);
 	}
+	
+	if (backup_crontab(n) == -1) {
+		fprintf(stderr, "no crontab for %s\n", User);
+		exit(ERROR_EXIT);
+	}
+
 	if (unlink(n) != 0) {
-		if (errno == ENOENT)
-			fprintf(stderr, "no crontab for %s\n", User);
-		else
-			perror(n);
+		perror(n);
 		exit(ERROR_EXIT);
 	}
 	poke_daemon();
 }
 
+/* returns	0	on success
+ *		-1	on non existent crontab file
+ *		-2	on failure to write the backup file
+ */
+static int backup_crontab(char crontab_filepath[]) {
+	const char *env_value;
+	char backup_filename[MAX_FNAME];
+	int ch = '\0';
+	FILE *crontab_file;
+	FILE *backup_file;
+	
+	if ((env_value = getenv("XDG_CACHE_HOME")) != NULL) {
+		if (!glue_strings(backup_filename, sizeof backup_filename, env_value,
+			"crontab.bak", '/')) {
+				fprintf(stderr, "path too long\n");
+				exit(ERROR_EXIT);
+		}
+	}
+	else if ((env_value = getenv("HOME")) != NULL) {
+		if (!glue_strings(backup_filename, sizeof backup_filename, env_value,
+			".cache/crontab.bak", '/')) {
+				fprintf(stderr, "path too long\n");
+				exit(ERROR_EXIT);
+		}
+	}
+	else {
+		fprintf(stderr, "Could not find environment variable XDG_CACHE_HOME or HOME to save the backup\n");
+		exit(ERROR_EXIT);
+	}
+
+	if ((crontab_file = fopen(crontab_filepath, "r")) == NULL) {
+		if (errno != ENOENT) {
+			perror(crontab_filepath);
+			exit(ERROR_EXIT);
+		}
+		return -1;
+	}
+	
+	if (swap_uids() == -1) {
+		perror("swapping uids");
+		exit(ERROR_EXIT);
+	}
+	if ((backup_file = fopen(backup_filename, "w+")) == NULL) {
+		perror(backup_filename);
+		fprintf(stderr, "Failed to write to the backup file\n");
+
+		if (swap_uids_back() < OK) {
+			perror("swapping uids back");
+			exit(ERROR_EXIT);
+		}
```

Comment:
To avoid duplication of this code, just split the backup_file assignment with the fopen() call and the check for backup_file == NULL please.